src/main/java/org/elasticsearch/index/engine/internal/InternalEngine.java

```diff
@@ -401,11 +401,11 @@ private void innerCreateNoLock(Create create, IndexWriter writer, long currentVe
         if ((versionValue != null && versionValue.delete() == false) || (versionValue == null && currentVersion != Versions.NOT_FOUND)) {
             if (create.origin() == Operation.Origin.RECOVERY) {
                 return;
-            } else if (create.origin() == Operation.Origin.REPLICA) {
```

Comment:
Without the change there are two scenarios:

1. First and second create request arrive in order on replica: on the replica 'doUpdate' is set but the version is not set to 1 -> versions on primary and replica are out of sync. 

2. First and second create request arrive out of order on primary: we get a 'DocumentAlreadyExistsException' because none of the other criteria match.

The tests I added in InternalEngineTests fail that way.
It seems to me we have to change something to make this work but I might be missing something.

@bleskes I agree we do not have to update the doc at all if we find it already exists and create.autoGeneratedId() is true. We can just return. I added a commit, test pass.